### PR TITLE
Pino 7 upgrade

### PR DESCRIPTION
**Pull request by Eomm**
Ref https://github.com/pinojs/pino/issues/1056

### CODE REVIEW

File: Readme.md [hunk 1]
```diff
@@ -70,7 +70,36 @@ $ echo '{"pid":94473,"hostname":"MacBook-Pro-3","level":30,"msg":"hello world","
...
-## Install
+
+## Usage as Pino Transport
+
+You can use this module as a [pino transport](https://getpino.io/#/docs/transports?id=v7-transports) like so:
+
+```js
+const pino = require('pino')
+const transport = pino.transport({
+  target: 'pino-toke',
```
Comment (jsumners) [inline Readme.md:81]: This seems incorrect.

File: app.js [hunk 2]
```diff
@@ -6,4 +6,9 @@ const fs = require('fs')
 const realPath = fs.realpathSync(__dirname)
 const script = path.join(realPath, 'psyslog.js')
 
-require(script.toString())
+module.exports = require('./lib/transport')
+
+if (require.main === module) {
+  // used as cli
+  require(script.toString())
+}
```
Comment (jsumners) [inline app.js:14]: Why this change? `app.js` is the "binary" script.

File: lib/rfc3164.js [hunk 3]
```diff
@@ -5,39 +5,8 @@ const stringify = require('fast-safe-stringify')
...
-
-  return message
-}
-
-module.exports = function ($options) {
-  options = $options
-
-  return through2.obj(function transport (data, enc, cb) {
+function messageBuilderFactory (options) {
+  return function (data) {
```
Comment (jsumners) [inline lib/rfc3164.js:9]: Can you please explain the changes here? They are really difficult to follow in a diff.

File: lib/rfc5424.js [hunk 4]
```diff
@@ -5,30 +5,9 @@ const stringify = require('fast-safe-stringify')
...
-  } else {
-    message = stringify(data)
-  }
-
-  return message
-}
-
-module.exports = function ($options) {
-  options = $options
+function messageBuilderFactory (options) {
```
Comment (jsumners) [inline lib/rfc5424.js:8]: Same as other spec file. Please explain.

**Review by jsumners — CHANGES_REQUESTED** [review]

File: app.js (hunk 2 above)
Comment (Eomm) [reply app.js:14]: The `"main"` was `"psyslog.js"` and I updated it to `app.js` (this file).
I should change it to `lib/transport`

File: Readme.md (hunk 1 above)
Comment (Eomm) [reply Readme.md:81]:
```suggestion
  target: 'pino-syslog',
```

File: lib/rfc3164.js (hunk 3 above)
Comment (Eomm) [reply lib/rfc3164.js:9]: I have extracted the logic the build the message string to write.
This logic is used in this file and in the `lib/transport` too.

The main changes are:

- removed the global `options` object (that would not work if the user set multiple transports with different configurations)
- exporting the build message function

File: lib/rfc5424.js (hunk 4 above)
Comment (Eomm) [reply lib/rfc5424.js:8]: This file has the same changes as the previous one:

- removed global option object that was shared across functions
- isolated the function to build the message string to write out
- expose the build message function

File: app.js (hunk 2 above)
Comment (jsumners) [reply app.js:14]: Yeah, the `main` was incorrect for programatic usage. It never mattered because this was designed for CLI usage only.

**Comment by jsumners** [comment]
Not sure it matters since this should be a new major, but tests are failing on 12.

**Comment by mcollina** [comment]
We are still supporting v12 in pino@7.

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by jsumners** [comment]
> We are still supporting v12 in pino@7.

Then tests should pass.

File: lib/transport.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  return build(function (source) {
+    source.on('data', function (obj) {
+      const canContinue = destination.write(processMessage(obj))
+      if (!canContinue) {
+        source.pause()
+        destination.once('drain', () => { source.resume() })
+      }
+    })
+  }, {
+    // close (err, cb) {
```
Comment (Eomm) [inline lib/transport.js:27]: It seems that the `.end` callback is never executed

Comment (jsumners) [reply lib/transport.js:27]: Why not?

Comment (Eomm) [reply lib/transport.js:27]: If I uncomment this section, the tests are failing

I need to track down why this is happening

**Review by mcollina — CHANGES_REQUESTED** [review]
I think this needs to embed pino-socket. This module is supposed to be used with it and transports cannot be piped together.

**Comment by jsumners** [comment]
They are independent modules. We don't know what people do with the socket module. It's really, _really_, generic and can be used for many things even outside of Pino logs. It should remain a standalone module. Instead, we should be able to pipe transports into one another.

**Comment by mcollina** [comment]
I meant to add that functionality here, not viceversa as this  module is supposed to work with pino-socket.
Anyway, I'll add the concept of a pipeline to the transport system.

File: lib/transport.js (hunk 5 above)
Comment (Eomm) [reply lib/transport.js]: The issues was cause by the `end` usage.
When the destination is a `sonic-boom` stream, the `end` method does not accept any callback
https://github.com/pinojs/sonic-boom#sonicboomend

Instead, when a writable stream is set (stdout or stderr) the `end` consumes the callback correctly

**Comment by Eomm** [comment]
> I think this needs to embed pino-socket. This module is supposed to be used with it and transports cannot be piped together.

I will try to integrate it

**Comment by Eomm** [comment]
Testing 
https://github.com/pinojs/pino-abstract-transport/pull/25
https://github.com/pinojs/pino/pull/1143

File: Readme.md [hunk 6]
```diff
@@ -70,7 +70,36 @@ $ echo '{"pid":94473,"hostname":"MacBook-Pro-3","level":30,"msg":"hello world","
...
+const transport = pino.transport({
+  target: 'pino-syslog',
+  level: 'info',
+  options: {
+    destination: 1, // optional (default stdout)
+    ... // other options
+  }
+})
+pino(transport)
+```
```
Comment (mcollina) [inline Readme.md:89]: Can you please update these docs with the usage of pino-socket as well?

Comment (jsumners) [reply Readme.md:89]: Do we need to directly depend on and integrate with `pino-sockect`? Shouldn't it just be a documentation example with maybe an included test?

Comment (Eomm) [reply Readme.md:89]: > Can you please update these docs with the usage of pino-socket as well?

Sure

> Do we need to directly depend on and integrate with pino-sockect? Shouldn't it just be a documentation example with maybe an included test?

Yes, it is. the `pino-socket` is a dev-dep and you should be able to pipe the output to this module to other destinations (such mongodb)

File: test/transport.test.js [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+      .filter(line => line) // skip empty lines
+      .forEach(line => {
+        t.ok(line.startsWith(expected.shift()))
+      })
+  })
+    .then((serverSocket) => {
+      t.teardown(() => {
+        serverSocket.close()
+        serverSocket.unref()
+        transport.end()
```
Comment (Eomm) [inline test/transport.test.js:119]: I had to add this line to make it works on node12.

I thought it was due the `duplexyfy` stream, that accepts these two parameters:

```
      autoDestroy: true,
      destroy: true
```

But it was not.
Nether the underlining `socket.unref()` was the issue.

I would like to understand why the transport doesn't close. 
I still don not figured it out.

Comment (mcollina) [reply test/transport.test.js:119]: What was the problem? why the test did not pass?

Comment (Eomm) [reply test/transport.test.js:119]: `tap` triggers a timeout error - locally within an infinity timeout, the process does not end without any errors

File: test/transport.test.js [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+          },
+          {
+            target: 'pino-socket',
+            options: {
+              mode: 'tcp',
+              address,
+              port
+            }
+          }
+        ]
```
Comment (jsumners) [inline test/transport.test.js:143]: What would happen if a third transport is added? Are we saying that a transport with pipelining enabled will always pipe to the immediate next tansport in a transports array?

Comment (mcollina) [reply test/transport.test.js:143]: Yes exactly. There is no fan-out, it's functionally equivalent to `cat a | grep something | cat -`.

**Comment by mcollina** [comment]
@Eomm pino@7.0.0-rc.8 is out, you can use update this and use `pipeline`.

**Comment by Eomm** [comment]
> @Eomm pino@7.0.0-rc.8 is out, you can use update this and use `pipeline`.

Done

File: Readme.md (hunk 6 above)
Comment (mcollina) [reply Readme.md:89]: I think the main usecase is to use `pipeline`.

File: Readme.md [hunk 9]
```diff
@@ -70,7 +70,66 @@ $ echo '{"pid":94473,"hostname":"MacBook-Pro-3","level":30,"msg":"hello world","
...
+transport pipeline setting the `enablePipelining` option. In this case, the `destination` option is ignored.
+
+```js
+const transport = pino.transport({
+  pipeline: [
+    {
+      target: 'pino-syslog',
+      level: 'info',
+      options: {
+        enablePipelining: true,
```
Comment (mcollina) [inline Readme.md:109]: I think this should be the default

Comment (Eomm) [reply Readme.md]: Done

**Comment by Eomm** [comment]
There is something else I can do on this PR?

**Comment by mcollina** [comment]
@jsumners anything left for you?
